### ISSUE TITLE
sql: improve AS OF SYSTEM TIME to use and accept HLCs

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#9859]
When looking into #9779, it became apparent that AS OF SYSTEM TIME needs two changes.

1) It should use the HLC instead of the context clock to determine the max timestamp (the you cannot run an AS OF query in the future). Using the context clock is problematic because `cluster_logical_timestamp()` can return values higher than it, and thus an immediate read at a previously ready value with the clt() function will fail. For example:

```
INSERT INTO t VALUES (1) RETURNING cluster_logical_timestamp()
```

Say this clt return is saved as `timestamp`. Then:

```
SELECT * FROM t AS OF SYSTEM TIME <clt>
```

Can in some cases trigger the max timestamp error.

2) Support the full HLC precision as returned by CLT: both the nanosecond and logical ticks parts. Not supporting the logical ticks means having to add a full nanosecond, which causes some weird fallout in the AsOfRetry test.

**Comment by tschottdorf** [comment]
As an aside, and maybe this is already taken care of, we must not allowed that user-specified timestamp to enter the HLC in the first place. That is, if the user specifies a HLC timestamp that is ahead of the current HLC timestamp at the time the request is processed, we should let it error out. It should not get a chance to *update* our HLC.

What is relatively simple to do is that if the supplied timestamp is indeed ahead of our physical clock, you simply sleep that goroutine until it isn't any more, and then go on.

That avoids anything fishy and also the penalty should be rare.

**Comment by mjibson** [comment]
@tschottdorf current we do this: https://github.com/cockroachdb/cockroach/blob/84169aa225d5718095f381a98c59e78acb399ad3/sql/executor.go#L1402

I think that's early enough to prevent this from happening.

**Comment by tschottdorf** [comment]
Yep, that looks early enough, thanks.

**Comment by mjibson** [comment]
Fixed by #9934.
